libs/utils/src/logging.rs

```diff
@@ -330,37 +330,89 @@ impl std::fmt::Debug for SecretString {
 ///
 /// TODO: consider upgrading this to a warning, but currently it fires too often.
 #[inline]
-pub async fn log_slow<O>(name: &str, threshold: Duration, f: impl Future<Output = O>) -> O {
-    // TODO: we unfortunately have to pin the future on the heap, since GetPage futures are huge and
-    // won't fit on the stack.
-    let mut f = Box::pin(f);
+pub async fn log_slow<F, O>(name: &str, threshold: Duration, f: std::pin::Pin<&mut F>) -> O
+where
+    F: Future<Output = O>,
+{
+    monitor_slow_future(
+        threshold,
+        threshold, // period = threshold
+        f,
+        |MonitorSlowFutureCallback {
+             ready,
+             elapsed_total,
+             elapsed_since_last_callback: _,
+         }| {
+            if ready {
+                // NB: we check if we exceeded the threshold even if the timeout never fired, because
+                // scheduling or execution delays may cause the future to succeed even if it exceeds the
+                // timeout. This costs an extra unconditional clock reading, but seems worth it to avoid
+                // false negatives.
```

Comment:
nit: the part about the additional clock reading no longer applies.

Reply:
Why does it no longer apply? I guess it's just in the wrong place. I'll remove it.